spec_util/hash_keys.go

```diff
@@ -0,0 +1,69 @@
+package spec_util
+
+import (
+	pb "github.com/akitasoftware/akita-ir/go/api_spec"
+	"github.com/pkg/errors"
+
+	"github.com/akitasoftware/akita-libs/pbhash"
+	"github.com/akitasoftware/akita-libs/visitors/go_ast"
+	"github.com/akitasoftware/akita-libs/visitors/http_rest"
+)
+
+type hashOneOfVisitor struct {
+	http_rest.DefaultHttpRestSpecVisitor
+	err error
+}
+
+func (vis *hashOneOfVisitor) VisitData(c http_rest.HttpRestSpecVisitorContext, p *pb.Data) bool {
+	oneOf := p.GetOneof()
+	if oneOf == nil {
+		return true
+	}
+	options := p.GetOneof().Options
+	for k, v := range options {
+		h, err := pbhash.HashProto(v)
+		if err != nil {
+			vis.err = err
+			return false
+		}
+		if k != h {
+			delete(options, k)
+			options[h] = v
```

Comment:
It's not a big problem, but doing it this way means you may visit the same option twice (and hash it twice.)  Or not, it's not guaranteed.

This could be avoided by extracting the keys from the map first, and iterating over the original set of keys.

Reply:
Indeed, mutating a data structure while you're iterating over it usually results in undesired behaviour.

Reply:
Thanks, fixed in 70b9fba.